geode-core/src/main/java/org/apache/geode/internal/cache/TXReservationMgr.java

```diff
@@ -66,28 +65,17 @@ public void releaseReservation(IdentityArrayList localLocks) {
   private void checkForConflict(TXRegionLockRequestImpl rr, IdentityArrayList localLocks)
       throws CommitConflictException {
     Object r = getRegionObject(rr);
-    Set keys = rr.getKeys();
-    Object oldValue = this.regionLocks.put(r, keys);
+    Map keys = rr.getKeys();
+    Map oldValue = this.regionLocks.put(r, keys);
     if (oldValue != null) {
       try {
         // we may have a conflict
-        Object[] keysArray = keys.toArray();
-        if (oldValue instanceof Set) {
-          checkSetForConflict(rr, (Set) oldValue, keysArray, localLocks);
-          IdentityArrayList newValue = new IdentityArrayList(2);
-          newValue.add(oldValue);
-          newValue.add(keys);
-          this.regionLocks.put(r, newValue);
-        } else {
-          IdentityArrayList al = (IdentityArrayList) oldValue;
-          int alSize = al.size();
-          Object[] alArray = al.getArrayRef();
-          for (int i = 0; i < alSize; i++) {
-            checkSetForConflict(rr, (Set) alArray[i], keysArray, localLocks);
-          }
-          al.add(keys);
-          this.regionLocks.put(r, al); // fix for bug 36689
-        }
+        checkSetForConflict(rr, oldValue, keys, localLocks);
+        HashMap<Object, Boolean> newValue =
+            new HashMap<Object, Boolean>(oldValue.size() + keys.size());
+        newValue.putAll(oldValue);
+        newValue.putAll(keys);
+        this.regionLocks.put(r, newValue);
```

Comment:
You may want to continue to use IdentityArrayList -- this would help when you release local locks. Each transaction would have its own set(map) of keys in the region, and only need to find the set hold by the current transaction to release (foundIt).